src/groups/group.jl

```diff
@@ -0,0 +1,326 @@
+
+The left-invariant canonical metric on a Lie group.
+"""
+struct LeftInvariantCanonicalMetric <: RiemannianMetric end
+
+"""
+    RightInvariantCanonicalMetric()
+
+The right-invariant canonical metric on a Lie group.
+"""
+struct RightInvariantCanonicalMetric <: RiemannianMetric end
+
+"""
+    BiInvariantCanonicalMetric()
+
+The bi-invariant canonical metric on a Lie group.
+"""
+struct BiInvariantCanonicalMetric <: RiemannianMetric end
+
+
+"""
+    ActionDirection
+
+Direction of action on a manifold, either [`LeftAction`](@ref) or
+[`RightAction`](@ref).
+"""
+abstract type ActionDirection end
+
+"""
+    LeftAction()
+
+Left action of a group on a manifold.
+"""
+struct LeftAction <: ActionDirection end
+
+"""
+    RightAction()
+
+Right action of a group on a manifold.
+"""
+struct RightAction <: ActionDirection end
+
+
+@doc doc"""
+    Identity(group::AbstractGroupManifold)
+
+The identity element of the group `group`.
+"""
+struct Identity{G<:AbstractGroupManifold}
+    group::G
+end
+
+(e::Identity)(x) = group_id(e.group, x)
+
+@doc doc"""
+    inv(G::AbstractGroupManifold, x)
+
+Inverse $x^{-1}$ of an element $x$, such that
+$x \cdot x^{-1} = x^{-1} \cdot x = e$.
+"""
+function inv!(G::AbstractGroupManifold, y, x)
+    error("inv not implemented on $(typeof(G)) for point $(typeof(x))")
+end
+
+function inv(G::AbstractGroupManifold, y, e::Identity)
+    indentity!(G, y)
+    return y
+end
+
+function inv(G::AbstractGroupManifold, x)
+    y = similar_result(G, inv, x)
+    inv!(G, y, x)
+    return y
+end
+
+inv(::AbstractGroupManifold, e::Identity) = e
+
+function group_id!(G::AbstractGroupManifold, x)
+    error("group_id! not implemented on $(typeof(G)) for point $(typeof(x))")
+end
+
+@doc doc"""
+    group_id(G::AbstractGroupManifold, x)
+
+Identity element $e$, such that for any element $x$,
+$x \cdot e = e \cdot x = x$. The returned element is of a similar type to `x`.
+"""
+function group_id(G::AbstractGroupManifold, x)
+    y = similar_result(G, inv, x)
+    group_id!(G, y, x)
+    return y
+end
+
+# Adapted from `afoldl` in `operators.jl` in Julia base.
+# expand recursively up to a point, then switch to a loop.
+_group_afoldl(op, G, a) = a
+_group_afoldl(op, G, a, b) = op(G,a,b)
+_group_afoldl(op, G, a, b, c...) = _group_afoldl(op, G, op(G, a, b), c...)
+
+function _group_afoldl(op,G,a,b,c,d,e,f,g,h,i,j,k,l,m,n,o,p,qs...)
+    y = op(G,op(G,op(G,op(G,op(G,op(G,op(G,op(G,op(G,op(G,op(G,op(G,op(G,op(G,op(G,a,b),c),d),e),f),g),h),i),j),k),l),m),n),o),p)
+    return op(G, y, reduce((x, y)->op(G, x, y), qs))
+end
+
+"""
+    compose_left(G::AbstractGroupManifold, xs...)
+
+Compose elements of `G` using their left translation upon each other.
+This composition is left-associative.
+"""
+compose_left(G::AbstractGroupManifold, xs...) = _group_afoldl(compose_left, G, xs...)
+compose_left(G::AbstractGroupManifold, x) = x
+function compose_left(G::AbstractGroupManifold, x, y)
+    z = similar_result(G, compose_left, x, y)
```

Comment:
How does this behave if both `x` and `y` are identity?

Reply:
Good catch, this case should have a separate method.

Reply:
Right, so many methods are needed here...